### PR TITLE
Add ReportIVTs compiler mode

### PR DESCRIPTION
**Pull request by 333fred**
This adds a new flag to the command-line compiler (and corresponding msbuild property) that reports on IVT information the compiler knows about and tries to help users figure out where IVT issues are coming from. Concretely, it changes the output in 3 ways:

1. After all regular diagnostics are calculated and emitted, we look for any diagnostics that are about inaccessible members. If we find any, and those members are from other assemblies, we print what assembly they're from.
2. After this, we print the current assembly identity, meaning name and public key.
3. Finally, we go through every reference passed to the compiler and print the assembly identity, wether the assembly has granted IVT to the assembly being compiled, and all IVTs the compiler sees this assembly as having.

Examples of what this output looks like are in the newly added commandline tests for the scenario.

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for reviews

**Review by jjonescz — APPROVED** [review]

**Review by genlu — APPROVED** [review]
IDE change lgtm

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs [hunk 1]
```diff
@@ -469,6 +469,12 @@ protected override Encoding StandardOutputEncoding
             get { return (string?)_store[nameof(LangVersion)]; }
         }
 
+        public bool ReportIVTs
```
Comment (jcouv) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:472]: Tagging @billwagner as FYI. This PR adds a new compiler option, which would need to be documented in [https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/) (and the VB counterpart)

File: src/Compilers/Core/Portable/MetadataReader/PEAssembly.cs [hunk 2]
```diff
@@ -169,6 +169,14 @@ internal IEnumerable<ImmutableArray<byte>> GetInternalsVisibleToPublicKeys(strin
             return result ?? SpecializedCollections.EmptyEnumerable<ImmutableArray<byte>>();
         }
 
+        internal IEnumerable<string> GetInternalsVisibleToAssemblyNames()
+        {
+            if (_lazyInternalsVisibleToMap == null)
+                Interlocked.CompareExchange(ref _lazyInternalsVisibleToMap, BuildInternalsVisibleToMap(), null);
```
Comment (jcouv) [inline src/Compilers/Core/Portable/MetadataReader/PEAssembly.cs:175]: Consider extracting this to a common method `EnsureInternalsVisibleToMapInitialized` or some such.

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs [hunk 3]
```diff
@@ -377,5 +377,35 @@ private protected override GeneratorDriver CreateGeneratorDriver(ParseOptions pa
         {
             return CSharpGeneratorDriver.Create(generators, additionalTexts, (CSharpParseOptions)parseOptions, analyzerConfigOptionsProvider);
         }
+
+        private protected override void DiagnoseBadAccesses(TextWriter consoleOutput, ErrorLogger? errorLogger, Compilation compilation, ImmutableArray<Diagnostic> diagnostics)
+        {
+            DiagnosticBag newDiagnostics = DiagnosticBag.GetInstance();
+            foreach (var diag in diagnostics)
+            {
+                var location = diag.Location;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs:386]: nit: consider inlining (only one usage)

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs [hunk 4]
```diff
@@ -377,5 +377,35 @@ private protected override GeneratorDriver CreateGeneratorDriver(ParseOptions pa
...
+
+                if (symbol is null || ReferenceEquals(compilation.Assembly, symbol.ContainingAssembly))
+                {
+                    // Can't be IVT related
+                    continue;
+                }
+
+                // '{0}' is defined in assembly '{1}'.
+                newDiagnostics.Add(new CSDiagnostic(
+                    new CSDiagnosticInfo(ErrorCode.ERR_SymbolDefinedInAssembly, symbol, symbol.ContainingAssembly),
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs:404]: Why not add this information unconditionally to the existing diagnostics (improve them instead of adding a secondary diagnostic)?

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs [hunk 5]
```diff
@@ -377,5 +377,35 @@ private protected override GeneratorDriver CreateGeneratorDriver(ParseOptions pa
...
+                    location));
+            }
+
+            ReportDiagnostics(newDiagnostics.ToReadOnlyAndFree(), consoleOutput, errorLogger, compilation);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs:408]: >newDiagnostics.ToReadOnlyAndFree() [](http://example.com/codeflow?start=30&length=34)

nit: we could avoid/reduce the allocation because `ReportDiagnostics` accepts an `IEnumerable`. Something like `ReportDiagnostics(localIterator(...), consoleOutput, errorLogger, compilation);`

File: src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs [hunk 6]
```diff
@@ -514,7 +514,19 @@ internal bool RuntimeSupportsCovariantReturnsOfClasses
         internal abstract ImmutableArray<AssemblySymbol> GetLinkedReferencedAssemblies();
         internal abstract void SetLinkedReferencedAssemblies(ImmutableArray<AssemblySymbol> assemblies);
 
+        IEnumerable<ImmutableArray<byte>> IAssemblySymbolInternal.GetInternalsVisibleToPublicKeys(string simpleName)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs:517]: Can we avoid adding private implementations by just making the existing abstract methods public?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs:517]: The existing patterns for these interface implementations is to have private implementations. I don't want to change that here.

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs (hunk 4 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs:404]: The existing diagnostics are spread out in a number of locations throughout the compiler, and I don't want to refactor them all to potentially report the existing message (in the case it can't possibly be related to an IVT) or an IVT-enhanced message. I don't think these messages are useful in most cases, only when you're trying to diagnose these issues, so I'd rather keep them separate with a post-processing step.

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs (hunk 5 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs:408]: I do not think avoiding an allocation is worth any effort here. These are not mainline scenarios.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs [hunk 7]
```diff
@@ -2119,6 +2119,16 @@ internal override IEnumerable<ImmutableArray<byte>> GetInternalsVisibleToPublicK
...
         }
 
+        internal override IEnumerable<string> GetInternalsVisibleToAssemblyNames()
+        {
+            EnsureAttributesAreBound();
+
+            if (_lazyInternalsVisibleToMap == null)
+                return SpecializedCollections.EmptyEnumerable<string>();
+
+            return _lazyInternalsVisibleToMap.Keys;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs:2129]: I'm not sure what ordering guarantees `Keys` provides. Do we care for determinism?

File: src/Compilers/CSharp/Test/CommandLine/CommandLineIVTTests.cs [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+
+using System.IO;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests;
+
+public class CommandLineIVTTests : CommandLineTestBase
+{
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineIVTTests.cs:13]: nit: extra blank line below

File: src/Compilers/VisualBasic/Test/CommandLine/CommandLineIVTTests.vb [hunk 9]
```diff
@@ -0,0 +1,514 @@
...
+    Assembly name: 'PresentationCore'
+    Public Keys:
+      0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9
+    Assembly name: 'PresentationFramework'
+    Public Keys:
+      0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9
+Assembly reference: 'N1, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'
+  Grants IVT to current assembly: False
+  Grants IVTs to:
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Test/CommandLine/CommandLineIVTTests.vb:91]: nit: consider omitting when empty

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 4)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs (hunk 7 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs:2129]: I'll order the way we loop through them. I don't _think_ it should actually matter for this case, but better to be safe.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 10]
```diff
@@ -599,6 +600,52 @@ private bool ReportDiagnostics(DiagnosticBag diagnostics, TextWriter consoleOutp
...
+            // Now, go through each of the referenced assemblies and print their IVT information.
+            foreach (var assembly in currentAssembly.Modules.First().ReferencedAssemblySymbols.OrderBy(a => a.Name))
+            {
+                // Assembly reference: '{0}'
+                //   Grants IVT to current assembly: {1}
+                //   Grants IVTs to:
+
+                var assemblyInternal = compilation.GetSymbolInternal<IAssemblySymbolInternal>(assembly);
+                bool grantsIvt = currentAssemblyInternal.AreInternalsVisibleToThisAssembly(assemblyInternal);
+                var assemblyPublicKey = AssemblyIdentity.PublicKeyToString(assembly.Identity.PublicKey);
```
[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:629]: It was, but isn't anymore. Thanks for the catch.

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 11]
```diff
@@ -1354,6 +1355,15 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string? ba
                                 AddDiagnostic(diagnostics, ErrorCode.ERR_StdInOptionProvidedButConsoleInputIsNotRedirected);
                             }
                             continue;
+
+                        case "reportivts":
+                        case "reportivts+":
+                            reportIVTs = true;
```
[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs:1361]: Doesn't seem useful to me.

**Comment by 333fred** [comment]
@jcouv addressed feedback.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 12]
```diff
@@ -1,5 +1,5 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<root>
+<>
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:2]: Is this change intentional?

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs:1361]: Did you test for `/reportivts:blah` or some such? I believe those should be ignored (that's what the check would be for)

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 7)

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 12 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:2]: Hmm, I think this was a end-tag extension fix up problem when there were merge conflicts.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 11)

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:472]: https://github.com/dotnet/docs/issues/35921 tracks documenting it.
